### ISSUE TITLE
Replace pure JS utf8 encoder/decoder with standard TextEncoder/Decoder

### ISSUE THREAD

**Issue by zslayton** [amzn/ion-js#618]
`decodeUtf8` and `encodeUtf8` are *much* slower than the native implementations offered by the now-ubiquitous [`TextEncoder`](https://developer.mozilla.org/en-US/docs/Web/API/TextEncoder) and [`TextDecoder`](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder) classes. We should migrate if our supported runtimes list will allow it.

See [this thread](https://github.com/amzn/ion-js/issues/616#issuecomment-644198579) for details.

**Comment by wesboyt** [comment]
I made the decision to use an unlicensed version of utf8decode because versions of node and browsers that we officially support/ed? did not include textdecoder.
For example on ubuntu 18.04 the latest version of node is 8x and does not include textdecoder. (Even though more recent distributions of 8 do so). The perf of the decode I yoinked is pretty terrible due to repeatedly concatenating strings.

Definetely needs to be replaced.

**Comment by desaikd** [comment]
According to an investigation I did on the `TextDecoder.decode` vs `makeReader` performance when passing `Uint8Array` as input to these methods,

I found out that after 512 bytes(threshold) of `Uint8Array` data `TextDecoder.decode` works faster than `makeReader`. but for data less than 512 bytes `makeReader` works faster. 

```
For more than 512 bytes of Uint8Array of data,

makeReader:
0.000015406160113891825
Binary Reader Large x 64,909 ops/sec ±4.30% (72 runs sampled)

TextDecoder.decode:
0.00001150648431228379
Binary Reader Large x 86,908 ops/sec ±2.48% (80 runs sampled)
```
From above it's clear that `TextDecoder.decode` works faster for more than 512 bytes of data.

```
For 512 or less than 512 bytes of Uint8Array of data,

makeReader:
0.000010301126157451875
Binary Reader Large x 97,077 ops/sec ±2.06% (81 runs sampled)

TextDecoder.decode:
0.0000114280178005657
Binary Reader Large x 87,504 ops/sec ±2.76% (74 runs sampled)
```

Here, taking into account the threshold of 512 bytes . It seems more reasonable to use the current implementation of `makeReader` for 512 or less than 512 bytes and `TextDecoder` method for more than 512 bytes. This gives us best of both worlds and improves performance when there is more than 512 bytes of data as input.

Also as not every platform would have runtime feature `TextDecoder`, we need to add a check to verify if global variable `TextDecoder` exist or not?
